mubench.pipeline/tasks/implementations/dataset_check.py

```diff
@@ -0,0 +1,213 @@
+import logging
+
+from typing import Dict, List, Optional, Any
+from os import listdir
+from os.path import join, isdir, basename, exists
+
+from data.misuse import Misuse
+from data.project import Project
+from data.project_version import ProjectVersion
+from data.snippets import get_snippets
+from tasks.project_version_misuse_task import ProjectVersionMisuseTask
+
+
+class DatasetCheck(ProjectVersionMisuseTask):
+    def __init__(self, datasets: Dict[str, List[str]], checkout_base_path: str, data_base_path: str):
+        super().__init__()
+        self.logger = logging.getLogger("datasetcheck")
+        self.datasets = datasets
+        self.checkout_base_path = checkout_base_path
+        self.data_base_path = data_base_path
+        self.misuses_not_listed_in_any_version = []
+
+    def start(self):
+        self.misuses_not_listed_in_any_version = self._get_all_misuses(self.data_base_path)
+
+    def process_project(self, project: Project):
+        self.logger = logging.getLogger("datasetcheck.project")
+        self._new_known_datasets_entry(project.id)
+        self._check_required_keys_in_project_yaml(project)
+        return super().process_project(project)
+
+    def process_project_version(self, project: Project, version: ProjectVersion):
+        self.logger = logging.getLogger("datasetcheck.project.version")
+        self._new_known_datasets_entry(version.id)
+        self._check_required_keys_in_version_yaml(project, version)
+        self._check_misuses_listed_in_version_exist(version)
+        return super().process_project_version(project, version)
+
+    def process_project_version_misuse(self, project: Project, version: ProjectVersion, misuse: Misuse):
+        self.logger = logging.getLogger("datasetcheck.project.version.misuse")
+        self._new_known_datasets_entry(misuse.id)
+        self._misuse_listed_in_version(misuse.id)
+        self._check_required_keys_in_misuse_yaml(project, version, misuse)
+        self._check_version_misuse_id_conflict(project.id, version.version_id, misuse.misuse_id)
```

Comment:
Since we create the ids as `project.version` or `project.misuse`, we may also have conflicts between a misuse and version that the misuse is not linked to.